### PR TITLE
allow empty object to be CEL value.

### LINKED ISSUE #120226: CEL: `has` macro wrongly returns false for an empty struct object
### What happened?

While migrating webhooks to CEL for one of the `oneof` structs in our CRD api, I noticed that the empty object in CEL is interpreted wrongly.

Empty map object seems to be interpreted correctly. In the example below, both map and object are sent empty (`{}`) in json output.

### What did you expect to happen?

`has(emptyObject)` returns true

### How can we reproduce it (as minimally and precisely as possible)?

json (captured on the client side) sent to the apiserver

```
I0829 09:13:54.107133  226332 request.go:1171] Request Body: {"apiVersion":"stas.statnett.no/v1beta2","kind":"Application","metadata":{"annotations":{"kubectl.kubernetes.io/last-applied-configuration":"{\"apiVersion\":\"v1beta2\",\"kind\":\"Application\",\"metadata\":{\"annotations\":{},\"name\":\"works-with-map\",\"namespace\":\"application-operator\"},\"spec\":{\"filesFrom\":[{\"mapDir\":{}},{\"emptyDir\":{}}],\"image\":\"test\",\"metadata\":{\"system\":\"stas\"}}}\n"},"name":"works-with-map","namespace":"application-operator"},"spec":{"filesFrom":[**{"mapDir":{}},{"emptyDir":{}}**],"image":"test","metadata":{"system":"stas"}}}
I0829 09:13:54.107227  226332 round_trippers.go:466] curl -v -XPOST  -H "Accept: application/json" -H "Content-Type: application/json" -H "User-Agent: kubectl/4.13.0 (linux/amd64) kubernetes/92b1a3d" 'https://k3d-application-operator-server-0:6443/apis/stas.statnett.no/v1beta2/namespaces/application-operator/applications?fieldManager=kubectl-client-side-apply&fieldValidation=Strict'
I0829 09:13:54.129367  226332 round_trippers.go:553] POST https://k3d-application-operator-server-0:6443/apis/v1beta2/namespaces/application-operator/applications?fieldManager=kubectl-client-side-apply&fieldValidation=Strict 422 Unprocessable Entity in 22 milliseconds
I0829 09:13:54.129443  226332 round_trippers.go:570] HTTP Statistics: GetConnection 0 ms ServerProcessing 21 ms Duration 22 ms
I0829 09:13:54.129453  226332 round_trippers.go:577] Response Headers:
I0829 09:13:54.129474  226332 round_trippers.go:580]     Content-Type: application/json
I0829 09:13:54.129483  226332 round_trippers.go:580]     X-Kubernetes-Pf-Flowschema-Uid: 10867c16-51c9-4f1a-af7f-24589b6a5cc1
I0829 09:13:54.129490  226332 round_trippers.go:580]     X-Kubernetes-Pf-Prioritylevel-Uid: 370cb2a3-36b4-463a-aaab-e3544be27037
I0829 09:13:54.129499  226332 round_trippers.go:580]     Content-Length: 768
I0829 09:13:54.129506  226332 round_trippers.go:580]     Date: Tue, 29 Aug 2023 07:13:54 GMT
I0829 09:13:54.129519  226332 round_trippers.go:580]     Audit-Id: 1b1e9c02-29a6-46e9-8ad8-e4ef12a52a45
I0829 09:13:54.129526  226332 round_trippers.go:580]     Cache-Control: no-cache, private
I0829 09:13:54.129579  226332 request.go:1171] Response Body: {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"Application \"works-with-map\" is invalid: spec.filesFrom[1]: Invalid value: \"object\": invalid object type, expected either Properties or AdditionalProperties with Allows=true and non-empty Schema evaluating rule: Must specify one and only one of {mapDir, emptyDir}","reason":"Invalid","details":{"name":"works-with-map","kind":"Application","causes":[{"reason":"FieldValueInvalid","message":"Invalid value: \"object\": invalid object type, expected either Properties or AdditionalProperties with Allows=true and non-empty Schema evaluating rule: Must specify one and only one of {mapDir, emptyDir}","field":"spec.filesFrom[1]"}]},"code":422}
The Application "works-with-map" is invalid: spec.filesFrom[1]: Invalid value: "object": invalid object type, expected either Properties or AdditionalProperties with Allows=true and non-empty Schema evaluating rule: Must specify one and only one of {mapDir, emptyDir}
```

CRD schema
```yaml
             filesFrom:
                description: List of ConfigMap / Secret / PersistentVolumeClaim /
                  EmptyDir that will be mounted into the container as files.
                items:
                  properties:
                    emptyDir:
                      type: object
                    mapDir:
                      additionalProperties:
                        type: string
                      type: object
                  type: object
                  x-kubernetes-validations:
                  - message: Must specify one and only one of {mapDir, emptyDir}
                    rule: '[has(self.mapDir), has(self.emptyDir)].exists_one(x,x)'
                maxItems: 512
                type: array
```

### Anything else we need to know?

_No response_

### Kubernetes version

<details>

```console
D:~/experiments/cel (⎈|k3d-application-operator) $ kubectl version
WARNING: This version information is deprecated and will be replaced with the output from kubectl version --short.  Use --output=yaml|json to get the full version.
Client Version: version.Info{Major:"1", Minor:"26", GitVersion:"v1.26.1", GitCommit:"92b1a3d0e5d092430b523f6541aa0c504b2222b3", GitTreeState:"clean", BuildDate:"2023-04-19T04:06:37Z", GoVersion:"go1.19.6", Compiler:"gc", Platform:"linux/amd64"}
Kustomize Version: v4.5.7
Server Version: version.Info{Major:"1", Minor:"27", GitVersion:"v1.27.4+k3s1", GitCommit:"36645e7311e9bdbbf2adb79ecd8bd68556bc86f6", GitTreeState:"clean", BuildDate:"2023-07-28T09:46:04Z", GoVersion:"go1.20.6", Compiler:"gc", Platform:"linux/amd64"}
```

</details>

### Cloud provider

<details>

</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
PRETTY_NAME="Ubuntu 22.04.3 LTS"
NAME="Ubuntu"
VERSION_ID="22.04"
VERSION="22.04.3 LTS (Jammy Jellyfish)"
VERSION_CODENAME=jammy
ID=ubuntu
ID_LIKE=debian
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
UBUNTU_CODENAME=jammy

$ uname -a
Linux devbox 5.15.90.1-microsoft-standard-WSL2 #1 SMP Fri Jan 27 02:56:13 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

### PR DESCRIPTION
**Pull request by jiahuif**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:
This PR handles the case where a openAPIV3Schema has neither Properties or AdditionalProperties set.
#### Which issue(s) this PR fixes:

Fixes #120226

#### Special notes for your reviewer:
This PR does not change how the value system handles unstructure maps (additionalProperties == true, and without schema), or the behavior of `x-kubernetes-preserve-unknown-fields`.

additionalProperties == false is not allowed in CRDs. [1]

#### Does this PR introduce a user-facing change?

```release-note
CEL can now correctly handle a CRD openAPIV3Schema that has neither Properties nor AdditionalProperties.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [1]: https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#validation
```

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/cel/common/values.go [hunk 1]
```diff
@@ -84,18 +84,22 @@ func UnstructuredToVal(unstructured interface{}, schema Schema) ref.Val {
...
-		if schema.IsXPreserveUnknownFields() {
-			return &unstructuredMap{
-				value:  m,
-				schema: schema,
-				propSchema: func(key string) (Schema, bool) {
-					return nil, false
-				},
-			}
+
+		// properties and additionalProperties are mutual exclusive, but nothing prevents the situation
```
Comment (cici37) [inline staging/src/k8s.io/apiserver/pkg/cel/common/values.go:88]: Do we need a test case for this?

**Comment by cici37** [comment]
/lgtm
/approve
let's cherry pick this fix to earlier releases. Thanks!

**Comment by cici37** [comment]
A test could potencially be added together with this case and for the cases: `additionalProperties: true` and `additionalProperties: false` values.
I'll leave it up to you to decide if add it in this PR or separate.
/hold

**Comment by cici37** [comment]
/lgtm
/assign @jpbetz 
for approval.

File: staging/src/k8s.io/apiserver/pkg/cel/common/values.go [hunk 2]
```diff
@@ -84,18 +84,22 @@ func UnstructuredToVal(unstructured interface{}, schema Schema) ref.Val {
...
+		// where both are missing.
+		// An object that (1) has no properties (2) has no additionalProperties or additionalProperties == false
+		// is treated as an empty object.
+		// An object that has additionalProperties == true or has x-preserve-unknown-field extension is treated as
+		// an unstructured map.
+		// Empty object vs unstructured map is differentiated by unstructuredMap implementation with the set schema.
+		// The resulting result remains the same.
```
Comment (jpbetz) [inline staging/src/k8s.io/apiserver/pkg/cel/common/values.go:95]: Retain in the documentation that this also handles the `x-kubernetes-preserve-unknown-fields` case?

Comment (jiahuif) [reply staging/src/k8s.io/apiserver/pkg/cel/common/values.go:95]: I've separated the original case in the comments. PTAL.

**Comment by shashankram** [comment]
> /lgtm /approve let's cherry pick this fix to earlier releases. Thanks!

@cici37 @jiahuif was this fix backported to prior releases?

**Pull request merged**